README.md
# Rcon Bypass
This is a simple module to provide security and functionality to the rcon system in SAPP, allowing
you to avoid boiler plate, providing an easier way to achieve communication between server and
client using the rcon system.

## How it works?

You can require this module on your code and submit different commands, for admins and for
interception purposes:
```lua
local rcon = require "rcon"

-- Override the rcon callback with the module callback
OnCommand = rcon.OnCommand

-- Provide a commands interceptor function
function rcon.commandInterceptor(playerIndex, command, environment, rconPassword)
    if (command == "reset") then
        -- Do stuff
    end
end

function OnScriptLoad()
    -- Attach rcon modifications to the game
    rcon.attach()

    -- Submit a safe to intecept rcon
    rcon.submitRcon("sled")

    -- Add your public and admin commands
    rcon.submitCommand("fly")
    rcon.submitCommand("taxi")

    rcon.submitAdminCommand("reset")
    rcon.submitAdminCommand("debugmode")
end

-- Do not forget to detach rcon at unloading script or when necessary
function OnScriptUnload()
    rcon.detach()
end
```

# Changelog
See this [markdown](CHANGELOG.md) for changelog details!


CHANGELOG.md
# 1.0.1

- Fixed a bug with commands being triggered twice when sending from rcon
- Added rcon environments to public module properties

# 1.0.0

- Initial release

rcon_bypass.lua
------------------------------------------------------------------------------
-- Rcon Bypass for SAPP
-- Author: Sledmine
-- Version: 1.0
-- Interceptor for specific rcon commands
------------------------------------------------------------------------------

local inspect = require "inspect"

api_version = "1.12.0.0"

-- Current rcon in the server
local serverRcon

-- Accepted rcon passwords
local acceptedRcons = {}
 
-- Commands to intercept
local commandsList = {}

-- Variables used to store patches address
local rcon_password_address
local rcon_command_failed_message_address

-- Internal functions
local function split(divider, string)
    if (divider == nil or divider == '') then return 1 end
    local position, array = 0, {}
    for st, sp in function() return string.find(string, divider, position, true) end do
        table.insert(array, string.sub(string, position, st-1))
        position = sp + 1
    end
    table.insert(array, string.sub(string, position))
    return array
end

local function arrayHas(array, value)
	value = string.gsub(value, "'", "")
	for k,v in pairs(array) do
		local wildcard = split(v, value)[1]
		if (v == value or wildcard == "") then
			return true
		end
	end
	return false
end

-- Public external functions
function submitRcon(rcon)
	if (not arrayHas(acceptedRcons, rcon)) then
		cprint("Adding new accepted rcon: " .. rcon)
		acceptedRcons[#acceptedRcons + 1] = rcon
		return true
	end
	return false
end

function submitCommand(command)
	if (not arrayHas(commandsList, command)) then
		cprint("Adding new intecepted command: " .. command)
		commandsList[#commandsList + 1] = command
		return true
	end
	return false
end

-- Internal logic functions
local function isCommandByPasseable(command)
	if (arrayHas(commandsList, command)) then
		return true
	end
	return false
end

local function isRconByPasseable(rcon)
	if (arrayHas(acceptedRcons, rcon)) then
		return true
	end
	return false
end

function OnScriptLoad()
	-- Get rcon patches address
	rcon_password_address = read_dword(sig_scan("7740BA??????008D9B000000008A01") + 0x3)
	rcon_command_failed_message_address = read_dword(sig_scan("B8????????E8??000000A1????????55") + 0x1)
	cprint("Rcon password address: " .. string.format("%x", rcon_password_address))
	cprint("Rcon failed message patch address: " .. string.format("%x", rcon_command_failed_message_address))

	if (rcon_command_failed_message_address and rcon_password_address) then
		-- Remove "rcon command failure" message to the player
        safe_write(true)
        write_byte(rcon_command_failed_message_address, 0)
		safe_write(false)
		
		-- Read current rcon in the server
		serverRcon = read_string(rcon_password_address)
		if (serverRcon) then
			cprint("Server rcon password is: '" .. tostring(serverRcon) .. "'")
			register_callback(cb['EVENT_COMMAND'], "onRcon")
		else
			cprint("ERROR!!!! At getting server rcon, please set and enable rcon on the server.")
		end
	else
		cprint("There was a problem obtaining rcon patches, please be sure you are using the correct SAPP version.")
    end
end

function onRcon(playerIndex, command, environment, interceptedRcon)
	if (environment == 1) then
		if (interceptedRcon == serverRcon) then
			return true
		elseif (isRconByPasseable(interceptedRcon)) then
			cprint("Intercepted rcon: " .. interceptedRcon)
			if (isCommandByPasseable(command)) then
				cprint("Intercepted command: " .. command)
				cprint("intercepted command size: " .. #command)
				
			else
				cprint("Intercepted command: " .. command .. " is not in the bypasseable commands.")
				say_all(get_var(playerIndex, "$name") .. " is sending illegal commands trough rcon, watch out.")
				--execute_command("sv_kick " .. playerIndex)
			end
			return false
		else
			say_all(get_var(playerIndex, "$name") .. " was kicked by sending wrong rcon password.")
			execute_command("sv_kick " .. playerIndex)
		end
	end
end

function OnScriptUnload()
	if (rcon_command_failed_message_address) then
		-- Restore patch to the original value
        safe_write(true)
        write_byte(rcon_command_failed_message_address, 0x72)
        safe_write(false)
    end
end

rcon.lua
------------------------------------------------------------------------------
-- Rcon Bypass
-- Sledmine
-- SAPP commands interceptor
------------------------------------------------------------------------------
local rcon = {}

-- Accepted rcon passwords
rcon.safeRcons = {}

-- Admin commands
rcon.adminCommands = {}

-- Commands to intercept
rcon.safeCommands = {}

rcon.commandInterceptor = nil

-- Internal functions
local function split(s, sep)
    if (sep == nil or sep == '') then return 1 end
    local position, array = 0, {}
    for st, sp in function() return string.find(s, sep, position, true) end do
        table.insert(array, string.sub(s, position, st-1))
        position = sp + 1
    end
    table.insert(array, string.sub(s, position))
    return array
end

local function listHas(list, value)
    value = string.gsub(value, "'", "")
    for k, element in pairs(list) do
        local wildcard = split(value, element)[1]
        if (element == value or wildcard == "") then
            return true
        end
    end
    return false
end

local function submit(list, value)
    if (not listHas(list, value)) then
        list[#list + 1] = value
        return true
    end
    return false
end

local function isABypassValue(list, value)
    if (listHas(list, value)) then
        return true
    end
    return false
end

local function isRconSafe(value)
    return isABypassValue(rcon.safeRcons, value)
end

local function isCommandSafe(value)
    return isABypassValue(rcon.safeCommands, value)
end

local function isAdminCommand(value)
    return isABypassValue(rcon.adminCommands, value)
end

-- Public functions and main usage

function rcon.submitRcon(rconValue)
    print("Adding new accepted rcon: " .. rconValue)
    return submit(rcon.safeRcons, rconValue)
end

function rcon.submitAdmimCommand(commandValue)
    print("Adding new accepted command: " .. commandValue)
    return submit(rcon.adminCommands, commandValue)
end

function rcon.submitCommand(commandValue)
    print("Adding new accepted command: " .. commandValue)
    return submit(rcon.safeCommands, commandValue)
end

function rcon.OnRcon(playerIndex, command, environment, interceptedRcon)
    if (environment == 1) then
        local playerName = get_var(playerIndex, "$name")
        if (interceptedRcon == rcon.serverRcon) then
            -- Normal rcon usage, allow command
            if (isAdminCommand(command)) then
                rcon.commandInterceptor(playerIndex, command, environment, interceptedRcon)
            end
            return true
        elseif (isRconSafe(interceptedRcon)) then
            -- This is an interceptable rcon command
            print("Intercepted rcon: " .. interceptedRcon)
            if (isCommandSafe(command)) then
                -- Rcon command it's an expected command, apply bypass
                print("Intercepted command: " .. command)
                print("intercepted command size: " .. #command)
                -- Execute interceptor
                rcon.commandInterceptor(playerIndex, command, environment, interceptedRcon)
            else
                print("Intercepted command: " .. command .. " is not in the bypasseable commands.")
                print(playerName .. " is sending illegal commands trough rcon, watch out!!!")
                say_all(playerName .. " is sending illegal commands trough rcon, watch out!!!")
                execute_command("sv_kick " .. playerIndex)
            end
            return false
        else
            say_all(playerName .. " was kicked by sending wrong rcon password.")
            execute_command("sv_kick " .. playerIndex)
        end
        return false
    end
end

function rcon.attach()
    if (server_type == "sapp") then
        rcon.passwordAddress = read_dword(sig_scan("7740BA??????008D9B000000008A01") + 0x3)
        rcon.failMessageAddress = read_dword(sig_scan("B8????????E8??000000A1????????55") + 0x1)
        if (rcon.passwordAddress and rcon.failMessageAddress) then
            -- Remove "rcon command failure" message
            safe_write(true)
            write_byte(rcon.failMessageAddress, 0x0)
            safe_write(false)
            -- Read current rcon in the server
            rcon.serverRcon = read_string(rcon.passwordAddress)
            if (rcon.serverRcon) then
                print("Server rcon password is: \"" .. rcon.serverRcon .. "\"")
            else
                print("Error, at getting server rcon, please set and enable rcon on the server.")
            end
        else
            print("Error, at obtaining rcon patches, please check SAPP version.")
        end
    end
end

function rcon.detach()
    if (rcon.failMessageAddress) then
        -- Restore "rcon command failure" message
        safe_write(true)
        write_byte(rcon.failMessageAddress, 0x72)
        safe_write(false)
    end
end

return rcon
